model/textparse/promparse_test.go

```diff
@@ -23,6 +24,45 @@ import (
 	"github.com/prometheus/prometheus/model/labels"
 )
 
+// lbls is a helper for the readability of the expectations.
+func typeAndUnitLabels(typeAndUnitEnabled bool, enabled, disabled labels.Labels) labels.Labels {
+	if typeAndUnitEnabled {
+		return enabled
+	}
+	return disabled
+}
+
+// todoDetectFamilySwitch exists because there's a known TODO that require dedicated PR and benchmarks for PROM-39.
+// OM and Prom text format do NOT require TYPE, HELP or UNIT lines. This means that metric families can switch without
+// those metadata entries e.g.:
+// ```
+// TYPE go_goroutines gauge
+// go_goroutines 33 # previous metric
+// different_metric_total 12 # <--- different family!
+// ```
+// The expected type for "different_metric_total" is obviously unknown type and unit, but it's surprisingly expensive and complex
+// to reliably write parser for those cases. Two main issues:
+// a. TYPE and UNIT are associated with "metric family" which is different than resulting metric name (e.g. histograms).
+// b. You have to alloc additional entries to pair TYPE and UNIT with metric families they refer to (nit)
```

Comment:
When starting a parser, could we create a pool of metric families with empty units and UNKNOWN metric types? I'm just spitting out ideas, I haven't tried anything myself 🤔

Reply:
Yea, it's all solvable just we we need to change a bit and spend time here, I was curious if we can finalize MVP first and create dedicated new PR for benchmarks...

For your question:
(a) is much bigger problem and (b) pool of metric families? Like every possible metric family names or what? (: Yea, it's only one extra string to store so "it's fine". We might be able to reduce some mem on scrape loop if we do this right in exchange.